File: internal/ns/menu_darwin.go
// Copyright ©2021 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package ns

import "github.com/progrium/macdriver/objc"

var menuClass = objc.Get("NSMenu")

// Menu https://developer.apple.com/documentation/appkit/nsmenu?language=objc
type Menu struct {
	objc.Object
}

// NewMenu https://developer.apple.com/documentation/appkit/nsmenu/1518144-initwithtitle?language=objc
func NewMenu(title string) Menu {
	titleStr := StringFromString(title)
	defer titleStr.Release()
	obj := menuClass.Alloc()
	obj.Send("initWithTitle:", titleStr)
	obj.Retain()
	return Menu{Object: obj}
}

// SetDelegate https://developer.apple.com/documentation/appkit/nsmenu/1518169-delegate?language=objc
func (m Menu) SetDelegate(delegate objc.Object) {
	m.Send("setDelegate:", delegate)
}

// NumberOfItems https://developer.apple.com/documentation/appkit/nsmenu/1518202-numberofitems?language=objc
func (m Menu) NumberOfItems() int {
	return int(m.Send("numberOfItems").Int())
}

// ItemAtIndex https://developer.apple.com/documentation/appkit/nsmenu/1518218-itematindex?language=objc
func (m Menu) ItemAtIndex(index int) MenuItem {
	return MenuItem{Object: m.Send("itemAtIndex:", index)}
}

// InsertItemAtIndex https://developer.apple.com/documentation/appkit/nsmenu/1518201-insertitem?language=objc
func (m Menu) InsertItemAtIndex(item MenuItem, index int) {
	m.Send("insertItem:atIndex:", item, index)
}

// RemoveItemAtIndex https://developer.apple.com/documentation/appkit/nsmenu/1518207-removeitematindex?language=objc
func (m Menu) RemoveItemAtIndex(index int) {
	m.Send("removeItemAtIndex:", index)
}

// Title https://developer.apple.com/documentation/appkit/nsmenu/1518192-title?language=objc
func (m Menu) Title() string {
	return m.Send("title").String()
}

File: internal/mac/menu_darwin.m
// Copyright (c) 2021-2026 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

#import "macos.h"

void goUpdateMenuCallback(NSMenuRef menu);

@interface MenuDelegate : NSObject<NSMenuDelegate>
@end

@implementation MenuDelegate
- (void)menuNeedsUpdate:(NSMenu *)menu {
	goUpdateMenuCallback((NSMenuRef)(menu));
}
@end

static MenuDelegate *menuDelegate = nil;

NSMenuRef newMenu(CFStringRef title) {
	NSMenu *menu = [[[NSMenu alloc] initWithTitle:(NSString *)title] retain];
	if (!menuDelegate) {
		menuDelegate = [MenuDelegate new];
	}
	[menu setDelegate:menuDelegate];
	return (NSMenuRef)menu;
}

int menuNumberOfItems(NSMenuRef m) {
	return [(NSMenu *)m numberOfItems];
}

NSMenuItemRef menuItemAtIndex(NSMenuRef m, int index) {
	return (NSMenuItemRef)[(NSMenu *)m itemAtIndex:index];
}

void menuInsertItemAtIndex(NSMenuRef m, NSMenuItemRef mi, int index) {
	[(NSMenu *)m insertItem:(NSMenuItem *)mi atIndex:index];
}

void menuRemoveItemAtIndex(NSMenuRef m, int index) {
	[(NSMenu *)m removeItemAtIndex:index];
}

void menuRemoveAll(NSMenuRef m) {
	[(NSMenu *)m removeAllItems];
}

CFStringRef menuTitle(NSMenuRef m) {
	return (CFStringRef)[(NSMenu *)m title];
}

void menuPopup(NSWindowRef wnd, NSMenuRef m, NSMenuItemRef mi, CGRect bounds) {
	// popupMenuPositioningItem:atLocation:inView: is not being used here because it fails to work when a modal dialog
	// is being used.
	NSPopUpButtonCell *popUpButtonCell = [[[NSPopUpButtonCell alloc] initTextCell:@"" pullsDown:NO] retain];
	[popUpButtonCell setAutoenablesItems:NO];
	[popUpButtonCell setAltersStateOfSelectedItem:NO];
	[popUpButtonCell setMenu:(NSMenu *)m];
	[popUpButtonCell selectItem:(NSMenuItem *)mi];
	[popUpButtonCell performClickWithFrame:bounds inView:[(NSWindow *)wnd contentView]];
	[popUpButtonCell release];
}
